sdk/core/Azure.Core.Expressions.DataFactory/src/DataFactoryExpressionJsonConverter.cs

```diff
@@ -2,41 +2,238 @@
+            if (typeToConvert == typeof(DataFactoryExpression<Array?>))
+                return Deserialize<Array>(document.RootElement);
+            if (typeToConvert == typeof(DataFactoryExpression<bool?>) || typeToConvert == typeof(DataFactoryExpression<bool>))
+                return Deserialize<bool>(document.RootElement);
+            if (IsGenericDataFactoryList(typeToConvert))
+            {
+                string methodName = typeToConvert.GenericTypeArguments[0].GetGenericTypeDefinition() == typeof(IList<>)
+                    ? nameof(DeserializeList)
+                    : nameof(DeserializeReadOnlyList);
+                var methodInfo = GetGenericSerializationMethod(typeToConvert, methodName);
+                return methodInfo!.Invoke(null, new object[] { document.RootElement })!;
+            }
+
             throw new InvalidOperationException($"Unable to convert {typeToConvert.Name} into a DataFactoryExpression<T>");
         }
 
-        public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+        public override void Write(Utf8JsonWriter writer, object? value, JsonSerializerOptions options)
+        {
+            if (value is null)
+                writer.WriteNullValue();
+            else if (value is DataFactoryExpression<string?> stringExpression)
+                Serialize(writer, stringExpression);
+            else if (value is DataFactoryExpression<int> intExpression)
+                Serialize(writer, intExpression);
+            else if (value is DataFactoryExpression<int?> nullableIntExpression)
+                Serialize(writer, nullableIntExpression);
+            else if (value is DataFactoryExpression<double> doubleExpression)
+                Serialize(writer, doubleExpression);
+            else if (value is DataFactoryExpression<double?> nullableDoubleExpression)
+                Serialize(writer, nullableDoubleExpression);
+            else if (value is DataFactoryExpression<Array?> arrayExpression)
+                Serialize(writer, arrayExpression);
+            else if (value is DataFactoryExpression<bool> boolExpression)
+                Serialize(writer, boolExpression);
+            else if (value is DataFactoryExpression<bool?> nullableBoolExpression)
+                Serialize(writer, nullableBoolExpression);
+            else
+            {
+                Type typeToConvert = value.GetType();
+                if (IsGenericDataFactoryList(typeToConvert))
+                {
+                    string methodName = typeToConvert.GenericTypeArguments[0].GetGenericTypeDefinition() == typeof(IList<>)
+                        ? nameof(SerializeList)
+                        : nameof(SerializeReadOnlyList);
+                    var methodInfo = GetGenericSerializationMethod(typeToConvert, methodName);
+                    methodInfo!.Invoke(null, new object[] { writer, value });
+                }
```

Comment:
Yeah I considered that initially, essentially just passing back the type via an out param. It seemed a little weird, but you are right it would make the overall if/else pattern cleaner.